### PR TITLE
chore(launch): code jobs runtime support handling

### PR DESCRIPTION
**Pull request by gtarpenning**
https://wandb.atlassian.net/browse/WB-15978
issue: 

create job
Screen Shot 2023-09-14 at 5 26 33 PM

run job -- error with bad python version
Screen Shot 2023-09-14 at 5 26 19 PM

fix:
create job
Screen Shot 2023-09-14 at 5 29 13 PM

run job -- success
Screen Shot 2023-09-14 at 5 29 18 PM

(also patch docker tag issue when multiple colons: [issue](https://wandb.atlassian.net/browse/WB-15868))

### CODE REVIEW

File: wandb/sdk/internal/job_builder.py [hunk 1]
```diff
@@ -234,7 +234,9 @@ def _build_artifact_job_source(
         else:
             full_program_relpath = program_relpath
         entrypoint = [
-            os.path.basename(sys.executable),
+            f"python{runtime.split('.')[0]}"
+            if runtime
+            else os.path.basename(sys.executable),
```
Comment (gtarpenning) [inline wandb/sdk/internal/job_builder.py:239]: there is always a runtime...

File: wandb/sdk/internal/job_builder.py [hunk 2]
```diff
@@ -233,10 +233,8 @@ def _build_artifact_job_source(
                 full_program_relpath = os.path.basename(program_relpath)
         else:
             full_program_relpath = program_relpath
-        entrypoint = [
-            os.path.basename(sys.executable),
-            full_program_relpath,
-        ]
+        python = f"python{'.'.join(runtime.split('.')[:2])}"
```
Comment (gtarpenning) [inline wandb/sdk/internal/job_builder.py:236]: never ever use the current `os` python

Comment (KyleGoyette) [reply wandb/sdk/internal/job_builder.py:236]: When we use our default python base images yes.

When users provide a base image, its less clear.

**Review by KyleGoyette — COMMENTED** [review]
Im confused about why these changes dont impact git sourced jobs.

File: wandb/sdk/internal/job_builder.py (hunk 2 above)
Comment (gtarpenning) [reply wandb/sdk/internal/job_builder.py:236]: lets just add an update from metadata step like repo

File: wandb/sdk/internal/job_builder.py [hunk 3]
```diff
@@ -259,18 +241,54 @@ def _build_image_job_source(
 
         raw_image_name = image_name
         if ":" in image_name:
-            raw_image_name, tag = image_name.split(":")
-            self._aliases += [tag]
+            tag = image_name.split(":")[-1]
+
+            # if tag looks properly formatted, assume its a tag
+            # regex: alphanumeric add "_" "-" "."
```
Comment (gtarpenning) [inline wandb/sdk/internal/job_builder.py:247]:
```suggestion
            # regex: alphanumeric and "_" "-" "."
```

File: wandb/sdk/internal/job_builder.py [hunk 4]
```diff
@@ -259,18 +241,54 @@ def _build_image_job_source(
...
         return source, name
 
+    def _make_job_name(self, input_str: str) -> str:
+        """Use job name from settings if provided, else use programatic name."""
+        if self._settings.job_name:
+            return self._settings.job_name
+
+        return make_artifact_name_safe(f"job-{input_str}")
+
+    def _get_entrypoint_and_notebook(
```
Comment (KyleGoyette) [inline wandb/sdk/internal/job_builder.py:266]: nit: This function doesn't update `self._is_notebook_run` does it need to return it? Is this for readability?

File: wandb/sdk/internal/job_builder.py (hunk 3 above)
Comment (KyleGoyette) [reply wandb/sdk/internal/job_builder.py:247]: This is all associated with the second ticket you added in here right?

**Review by KyleGoyette — APPROVED** [review]
Approved with a question

**Pull request merged**